src/core/linalg/src/dense/4C_linalg_utils_densematrix_funct.cpp

```diff
@@ -0,0 +1,1769 @@
+  {
+    if (id_minus_A.norm2() < 1.0)
+      calc_method = Core::LinAlg::GenMatrixLogFirstDerivCalcMethod::taylor_series;
+    else
+      calc_method = Core::LinAlg::GenMatrixLogFirstDerivCalcMethod::gregory_series;
+  }
+
+  // compute derivative
+  switch (calc_method)
+  {
+    case GenMatrixLogFirstDerivCalcMethod::pade_part_fract:
+      FOUR_C_THROW(
+          "For the computation methods based on the Pade approximation, you have to specify the "
+          "Pade order for the computation as either input or output!");
+    case FourC::Core::LinAlg::GenMatrixLogFirstDerivCalcMethod::taylor_series:
+      return matrix_3x3_log_1st_deriv_taylor_series(input, err_status, 1.0e-10);
+    case GenMatrixLogFirstDerivCalcMethod::gregory_series:
+      return matrix_3x3_log_1st_deriv_gregory_series(input, err_status, 1.0e-10);
+    default:
+      FOUR_C_THROW(
+          "The provided computation method for first derivative of the matrix logarithm of a "
+          "general matrix is not implemented!");
+  }
+}
+
+/*--------------------------------------------------------------------*
+ *--------------------------------------------------------------------*/
+Core::LinAlg::Matrix<9, 9> Core::LinAlg::matrix_3x3_log_1st_deriv(
+    const Core::LinAlg::Matrix<3, 3>& input, Core::LinAlg::MatrixFunctErrorType& err_status,
+    const unsigned int pade_order, Core::LinAlg::GenMatrixLogFirstDerivCalcMethod calc_method)
+{
+  // auxiliaries
+  Core::LinAlg::Matrix<3, 3> id_3x3(true);
+  for (int i = 0; i < 3; ++i)
+  {
+    id_3x3(i, i) = 1.0;
+  }
+
+  // characteristic matrix \f$ \boldsymbol{I} - \boldsymbol{A} \f$
+  Core::LinAlg::Matrix<3, 3> id_minus_A(true);
+  id_minus_A.update(1.0, id_3x3, -1.0, input, 0.0);
+
+  // determine the computation method based on matrix characteristics
+  if (calc_method == Core::LinAlg::GenMatrixLogFirstDerivCalcMethod::default_series)
+  {
+    if (id_minus_A.norm2() < 1.0)
+      calc_method = Core::LinAlg::GenMatrixLogFirstDerivCalcMethod::taylor_series;
+    else
+      calc_method = Core::LinAlg::GenMatrixLogFirstDerivCalcMethod::gregory_series;
+  }
+
+  // compute derivative
+  switch (calc_method)
+  {
+    case GenMatrixLogFirstDerivCalcMethod::pade_part_fract:
+      return matrix_3x3_log_1st_deriv_pade_part_fract(input, err_status, pade_order);
+    default:
+      return matrix_3x3_log_1st_deriv(input, err_status);
+  }
+}
+
+
+
+/*--------------------------------------------------------------------*
+ *--------------------------------------------------------------------*/
+void Core::LinAlg::sym_matrix_3x3_exp_2nd_deriv_voigt(const Core::LinAlg::Matrix<3, 3>& input,
+    Core::LinAlg::Matrix<3, 3>& exp, Core::LinAlg::Matrix<6, 6>& dexp_mat,
+    Core::LinAlg::Matrix<6, 6>* ddexp_mat, Core::LinAlg::MatrixFunctErrorType& err_status)
+{
+  // declare arrays of derivatives
+  Core::LinAlg::Matrix<3, 3> matrix_exp_1st_deriv[6];
```

Comment:
Why don't you use a std::array?
```suggestion
  std::array<Core::LinAlg::Matrix<3, 3>,6> matrix_exp_1st_deriv;
```